### ISSUE TITLE
Possible error check missing in rayCast ?

### ISSUE THREAD

**Issue by icoderaven** [NVIDIA/gvdb-voxels#79]
I've been struggling for a bit on trying to isolate the cause of a repeatable indeterministic error that pops up in my GVDB pipeline. The error seems to arise in the inner loop of the 3DDA algorithm right before calling the brickFunc. Specifically, the nodeid I get is a spurious value, 
```
(cuda-gdb) p nodeid
$67 = {-589654601, 24, 0, -1090300387, 406}
```
and the node pointed to seems to be nonsensical
```
(cuda-gdb) p *node
$68 = {mLev = 0 '\000', mFlags = 0 '\000', mPriority = 0 '\000', pad = 0 '\000', 
  mPos = {x = 0, y = 0, z = 0}, mValue = {x = 0, y = 0, z = 0}, mVRange = {x = 0, 
    y = 0, z = 0}, mParent = 0, mChildList = 0, mMask = 0}
(cuda-gdb) p b
$70 = 512
(cuda-gdb) p p
$71 = {x = 8, y = 7, z = 7}
```
I tried to cache the previous node, and it doesn't seem to provide anything of much help, except that its parent is ID_UNDEFL, so I'm assuming it's a root node with a {-1,-1,-1} mValue. That looks very wrong to me.
```
(cuda-gdb) p *prev_node
$69 = {mLev = 2 '\002', mFlags = 1 '\001', mPriority = 12 '\f', pad = 94 '^', 
  mPos = {x = 0, y = 0, z = 0}, mValue = {x = -1, y = -1, z = -1}, mVRange = {
    x = 1.55819629e+28, y = 1.94444439e+31, z = 1.53989704e+10}, 
  mParent = 4294967295, mChildList = 513, mMask = 7310299209757123950}
```
In my application, I obtain depth images that I clip within bounds and then add to the volume after scaling them up. Each iteration looks something like this:
```
  DataPtr imgptr, temp1, temp2;
  uint num_points = points.rows();
  gvdb_.AllocData(imgptr, num_points, sizeof(Vector3DF), false);
  gvdb_.CommitData(imgptr, num_points, (char *)points.data(), 0, sizeof(Vector3DF));
  gvdb_.SetPoints(imgptr, temp1, temp2);
  Vector3DF m_origin = Vector3DF(0, 0, 0);
  float m_radius = 1.0f;
  gvdb_.AccumulateTopology(num_points, m_radius, m_origin);
  gvdb_.FinishTopology(true, true);
  gvdb_.UpdateAtlas();
  if (imgptr.gpu != 0x0) {
    cudaCheck(cuMemFree(imgptr.gpu), "GVDBMRFMap", "activateSpace", "cuMemFree", "", false);
  }
```
Each depth image is around 848x480 pixels. I set the map resolution to be 0.01 (so a multiplier of 100 for depth ranges in 0-5m. GVDB configuration of <1,3,3,3>. Incoming data clipped in a 2x2x2m cube. So effective resolution 2/0.01 ~= 200 voxels). The only other modification to this pathway I have in the library is that I increased the temporary size of the buffer in ActivateIncreBricksGPU
```
PrepareAux ( AUX_BRICK_LEVXYZ, pNumPnts * pRootLev * 4 * 8 / Incre_ratio, sizeof(unsigned short), true );
```
What gives? Am I missing something? Or is there a sneaky bug in the list generation?

Every 19th image I get a kernel error. An interesting observation was that if I called my codebase from my pybind-ed codebase, I saw this error crop up earlier. My initial suspicion was the GPU Context being garbage collected, but I think I have ruled that out via stepping through cuda-gdb.

I noticed that there was a new giant commit pushed, but wanted to know where best to tackle this right now as it'll take me time to reconcile my changes with the new upstream changes, and I would really like to use what I have since it *almost* works right. Thanks!!

**Comment by NeilBickford-NV** [comment]
Hi icoderaven,

Thanks for making an issue for this - sorry for taking so long to get to this! This is very weird - e.g. `nodeid` should be an int in `cuda_gvdb_raycast.cuh`, but in the first sample `cuda-gdb` seems to be printing it as a 5-tuple! Have you managed to figure out what was going on in the meantime, or a sample I can maybe look at to try to narrow things down?

Thanks!

**Comment by icoderaven** [comment]
Hi @NeilBickford-NV ! Thanks for looking at this (and the project - really nice to not to have to worry about cudpp anymore)! I did manage to merge all your latest upstream changes with my fork without too many incidents, and I think I still have the issue, but I have a local hack that just aborts the raytrace when I face this issue for now. I'm currently working on a paper submission but should be able to reinvestigate soon!
p.s. The only major thing I'd like to change in your build refactor is to bring back the generated PTX macros since it makes it much easier to incorporate the library in third party applications, and ensuring that the PTX are generated with the most relevant gpu arch flags (I can make a pull request if required based off of
https://github.com/icoderaven/gvdb-voxels/commit/8e0cf0086e2c983607b458dafe744d9689a33c6b
and 
https://github.com/icoderaven/gvdb-voxels/commit/e5d9e657f55d2be03d0bacec61e9f72917f9005d)

**Comment by NeilBickford-NV** [comment]
Thanks! I really like using `CUDA_GVDB_COPYDATA_PTX` and `CUDA_GVDB_MODULE_PTX` instead of `"cuda_gvdb_copydata.ptx"` and `"cuda_gvdb_module.ptx"` - is it OK if I go ahead and implement a different version of that, since there are some things that I might implement slightly differently? (My only worry about the generated PTX architecture change is if someone builds PTX on a system with a Turing GPU, for instance, that also needs to run on an older GPU, e.g. Pascal.)

**Comment by icoderaven** [comment]
Hi! Sorry for the tardy reply, yes please, go ahead and I'll happily merge it within my branch.

**Comment by NeilBickford-NV** [comment]
OK, I just pushed a commit that added some changes based off of [icoderaven@8e0cf00](https://github.com/icoderaven/gvdb-voxels/commit/8e0cf0086e2c983607b458dafe744d9689a33c6b)!

**Comment by icoderaven** [comment]
Closing this since I no longer have this issue after merging the 1.1.1 refactor into my codebase. I also had to fix AccumulateTopology to deal with the situation when adding a new set of points would increase the spanning cover and requiring a new root node to be created.
(See https://github.com/icoderaven/gvdb-voxels/commit/b3256a921deed84995ab74f1bd726616a1d804e7) - I can create a pull request for this if required.
